### ISSUE TITLE
Allow command substitution with `command` builtin

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5575]
There was a brief discussion on a commit for this a while back but I wanted to formally file this.

Fish 3.0 supports `command $var` in lieu of `eval $var`, so there's no longer a good reason not to allow `command (subexpr)`  as well.

**Comment by faho** [comment]
Where would the output go?

**Comment by mqudsi** [comment]
No, I mean literally executing `command (echo $pager)` (as ridiculously contrived as that example is).

**Comment by mqudsi** [comment]
This should also include supporting `builtin (echo builtin_name)`, as `builtin $builtin_name` is now allowed.

**Comment by floam** [comment]
I think this is a little different, because we made the choice to allow variables in the command position. We haven't made that decision for command substitutions. Technically these are decorators internally, with the `$foo` in` `command $foo` in the command position.

**Comment by mqudsi** [comment]
There's no difference between the two. A command substitution is specified as evaluating to the same type of token that a variable evaluates to.

**Comment by faho** [comment]
>A command substitution is specified as evaluating to the same type of token

Quite simply: It's not. There is no such specification, and we currently make this distinction.

**Comment by mqudsi** [comment]
Ok, wrong choice of words. A command substitution expands to the same type of token that a variable expands to (namely, an array of strings).

**Comment by kpengboy** [comment]
+1 to this feature request.

I am currently working on a project with a directory setup where all binary output goes in a `bin` directory at the project root. So `/path/to/project/foo/bar/baz.cc`, for instance, compiles to `/path/to/project/bin/foo/bar/baz`.

I've currently written a function, let's call it `bindir`, that, given the CWD, translates it to the corresponding binary directory. For instance, if the current CWD is `/path/to/project/foo/bar`, the `bindir` function outputs the string `/path/to/project/bin/foo/bar`.

Ideally, I would like to have a workflow like this:
```
/path/to/project/foo/bar> vi baz.cc  # Write some code
/path/to/project/foo/bar> make
/path/to/project/foo/bar> (bindir)/baz --arg1 --arg2  # Test out code
/path/to/project/foo/bar> vi baz.cc  # Make some changes
/path/to/project/foo/bar> make
/path/to/project/foo/bar> (bindir)/baz --arg1 --arg2  # Test it again
...
```

Currently, the `(bindir)/baz` command doesn't actually work because command substitution is not allowed in this case. For me, this is a regression compared to bash or zsh, which happily allow me to type `` `bindir`/baz``.

Now that we support having variables in commands, I suppose it would make sense to allow command substitution as well, subject to the same restrictions as the former?

**Comment by offbyone** [comment]
This also really helps for users of homebrew/linuxbrew, as one idiom these have is locating the unexported commands via `brew --prefix`:

```
# in bash
$ $(brew --prefix awscli)/libexec/bin/pip install awscli_keyring

# in fish
$ set PREFIX (brew --prefix awscli)
$ command $PREFIX/libexec/bin/pip install awscli_keyring
```

Fish is much more cumbersome here as we need the local var. What I really want is the same-ish thing as bash/zsh:

```
$ command (brew --prefix awscli)/libexec/bin/pip install awscli_keyring
```

**Comment by fregante** [comment]
I arrived here trying to run:

```fish
DIR=(pwd)
fish: Command substitutions not allowed
```

But this is allowed:

```fish
set DIR (pwd)
echo $DIR
```

Might help someone.

**Comment by henryiii** [comment]
Also very useful for pyenv, where you may want to run `(pyenv prefix 3.7.8)/bin/python` for example.

**Comment by faho** [comment]
Fun fact, this works:

```fish
thing=(echo foo) $thing
```

e.g.

```fish
python=(pyenv prefix 3.7.8)/bin/python $python
```

**Comment by kesarling** [comment]
Uh, its really annoying when I want to use `${ps -ef | grep msedge | head -1}` as a variable for one single command. Please allow command substitution.

**Comment by eguiraud** [comment]
Another valid usecase is executing a command from history, e.g. after selecting it from a list with fzf:

```
(history | fzf)
```

**Comment by theotheroracle** [comment]
is this issue concerning this ?
```
[I] hive@uranus : ~/Downloads ඞ
> (which ssh)
fish: command substitutions not allowed here
```

**Comment by zanchey** [comment]
> is this issue concerning this ?

yep

**Comment by theotheroracle** [comment]
another usecase is aliasing ssh to `kitty +kitten ssh` but occasionally when that breaks, to run the previous command, would be nice

**Comment by IsaacOscar** [comment]
I also found the lack of this feature annoying, my usecase was simply `(status dirname)/some-program args...` (i.e. execute the filed called `some-program` that is in the directory of the current script).

I've defined a function that seems to work perfectly though:
```
function do
    $argv[2] $argv[2..]
end
```
So you can do `do (status dirname)/some-program args...`.
This can execute builtins, if you don't want that then you can do `do command (status dirname)/some-program args...`.

So the lack of this feature isn't too annoying

**Comment by flowchartsman** [comment]
> Fun fact, this works: `thing=(echo foo) $thing`
This works, but it's a workaround.  This is common enough that it would be nice to have it supported.  For example, the gcloud CLI [explicitly asks users to do something similar](https://cloud.google.com/iap/docs/using-tcp-forwarding#increasing_the_tcp_upload_bandwidth) when running the version of python it installs:

`$(gcloud info --format="value(basic.python_location)") -m pip install <some module>`

I'm sure Fish has its reasons for not allowing this, however a builtin helper would be nice.

**Comment by krobelus** [comment]
> $(gcloud info --format="value(basic.python_location)") -m pip install <some module>

We should probably support this. My one concern is that it kills the other possible meaning:

	( foo; bar )

could be made to mean the same thing as

	begin; foo; bar; end

Could do this only for `()` and use the other behavior for `$()`, for superficial compat with other shells..
Continuing on this road I'd make `cd` scoped to command substitutions. Not sure. Might be too much magic in both that prevents users from understanding.

> As a Fish user, normally I see commands like this and I just remove the $

no need, $ works fine

> I would need to search yet more to find out what that reason is, and I'm already frustrated from having to search to get this far.

A quick git-blame shows it's been this way since the initial Git commit from 2005.

> Even a message like
> 
> fish: command substitutions not allowed here. Try var=(command) $var

yeah we can do that for the time being. Small chance of name collision but that's ok.

**Comment by ilyagr** [comment]
A minor problem with `var=(git rev-parse --show-toplevel)/scripts/something $var` is that there is no tab expansion if you type `var=(git rev-parse --show-toplevel)/<TAB>`.

There *is* tab expansion when typing `env (git rev-parse --show-toplevel)/<TAB>`. So, I recommend

    env (git rev-parse --show-toplevel)/scripts/something

on POSIXy systems

**Comment by jfxg** [comment]
Just sharing that this issue is annoying enough that I'm actually no longer using fish, despite liking it as I started using it over the last few months. Using substitutions to generate commands is fairly common (as noted by plenty of other comments in this _years_ long open issue), and its very common in my environment at work. It's become too frustrating for me to have to work around it.

It would be nice to at least have an official stance provided on the current status: Is this planned to be fixed at all? Is there a technical reason it hasn't been fixed, or are the devs opposed to fixing it for other reasons?
